osquery/tables/system/ssh_keys.cpp

```diff
@@ -15,12 +15,115 @@
 #include <osquery/filesystem/filesystem.h>
 #include <osquery/logger/logger.h>
 #include <osquery/tables/system/system_utils.h>
+#include <osquery/utils/scope_guard.h>
 #include <osquery/utils/system/system.h>
 
+#include <boost/algorithm/string.hpp>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+
 namespace osquery {
 namespace tables {
 
-const std::string kSSHUserKeysDir = ".ssh/";
+extern const std::string kSSHUserKeysDir = ".ssh";
+
+const std::string kOpenSshHeader = "-----BEGIN OPENSSH PRIVATE KEY-----";
+
+// The first bytes of an OpenSSH key are |key type|length of cipher name|cipher
+// name| so all unencrypted ed25519 keys should start with the value below
+// (encoded in base64) This magic string is (hex-ified):
+//  6f 70 65 6e 73 73 68 2d 6b 65 79 2d 76 31 00 00 00 00 04 6e 6f 6e 65
+// |--------------------------------------------|-----------|----------|
+// | o  p  e  n  s  s  h  -  k  e  y  -  v  1 \0|          4| n  o  n e|
+const std::string kOpenSshUnencryptedPrefix = "b3BlbnNzaC1rZXktdjEAAAAABG5vbmU";
+
+// openssl can't currently parse the new openssh keys
+bool isOpenSSHKey(std::string& keys_content) {
+  return boost::starts_with(keys_content, kOpenSshHeader);
+}
+
+// `true` if the openssh key is encrypted, `false` otherwise.
+bool isOpenSSHKeyEncrypted(std::string& keys_content) {
+  const std::string prefix = keys_content.substr(
+      kOpenSshHeader.size() + 1, kOpenSshUnencryptedPrefix.size());
+  return prefix != kOpenSshUnencryptedPrefix;
+}
+
+// parsePrivateKey returns true iff the key is valid.
+// Tries to parse the .PEM using openssl. If that fails, it checks
+// if it's an openssh key.
+bool parsePrivateKey(std::string& keys_content,
+                     int* key_type,
```

Comment:
Instead of passing pointers, can you pass the references? (essentially the same thing)

Reply:
Done